{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "QA6Ml_U3vbDX",
        "outputId": "a36fb8fa-9316-4d3b-e7dc-a10504b2dc58"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: dash in /usr/local/lib/python3.11/dist-packages (3.1.1)\n",
            "Requirement already satisfied: plotly in /usr/local/lib/python3.11/dist-packages (5.24.1)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (2.2.2)\n",
            "Requirement already satisfied: biopython in /usr/local/lib/python3.11/dist-packages (1.85)\n",
            "Requirement already satisfied: dash-bio in /usr/local/lib/python3.11/dist-packages (1.0.2)\n",
            "Requirement already satisfied: Flask<3.2,>=1.0.4 in /usr/local/lib/python3.11/dist-packages (from dash) (3.1.1)\n",
            "Requirement already satisfied: Werkzeug<3.2 in /usr/local/lib/python3.11/dist-packages (from dash) (3.1.3)\n",
            "Requirement already satisfied: importlib-metadata in /usr/local/lib/python3.11/dist-packages (from dash) (8.7.0)\n",
            "Requirement already satisfied: typing-extensions>=4.1.1 in /usr/local/lib/python3.11/dist-packages (from dash) (4.14.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from dash) (2.32.3)\n",
            "Requirement already satisfied: retrying in /usr/local/lib/python3.11/dist-packages (from dash) (1.4.0)\n",
            "Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.11/dist-packages (from dash) (1.6.0)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.11/dist-packages (from dash) (75.2.0)\n",
            "Requirement already satisfied: tenacity>=6.2.0 in /usr/local/lib/python3.11/dist-packages (from plotly) (8.5.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.11/dist-packages (from plotly) (24.2)\n",
            "Requirement already satisfied: numpy>=1.23.2 in /usr/local/lib/python3.11/dist-packages (from pandas) (2.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.11/dist-packages (from dash-bio) (1.15.3)\n",
            "Requirement already satisfied: scikit-learn>=0.20.1 in /usr/local/lib/python3.11/dist-packages (from dash-bio) (1.6.1)\n",
            "Requirement already satisfied: colour in /usr/local/lib/python3.11/dist-packages (from dash-bio) (0.1.5)\n",
            "Requirement already satisfied: GEOparse>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from dash-bio) (2.0.4)\n",
            "Requirement already satisfied: jsonschema in /usr/local/lib/python3.11/dist-packages (from dash-bio) (4.24.0)\n",
            "Requirement already satisfied: parmed in /usr/local/lib/python3.11/dist-packages (from dash-bio) (4.3.0)\n",
            "Requirement already satisfied: periodictable in /usr/local/lib/python3.11/dist-packages (from dash-bio) (2.0.2)\n",
            "Requirement already satisfied: blinker>=1.9.0 in /usr/local/lib/python3.11/dist-packages (from Flask<3.2,>=1.0.4->dash) (1.9.0)\n",
            "Requirement already satisfied: click>=8.1.3 in /usr/local/lib/python3.11/dist-packages (from Flask<3.2,>=1.0.4->dash) (8.2.1)\n",
            "Requirement already satisfied: itsdangerous>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from Flask<3.2,>=1.0.4->dash) (2.2.0)\n",
            "Requirement already satisfied: jinja2>=3.1.2 in /usr/local/lib/python3.11/dist-packages (from Flask<3.2,>=1.0.4->dash) (3.1.6)\n",
            "Requirement already satisfied: markupsafe>=2.1.1 in /usr/local/lib/python3.11/dist-packages (from Flask<3.2,>=1.0.4->dash) (3.0.2)\n",
            "Requirement already satisfied: tqdm>=4.31.1 in /usr/local/lib/python3.11/dist-packages (from GEOparse>=1.1.0->dash-bio) (4.67.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->dash) (3.4.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->dash) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->dash) (2.4.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->dash) (2025.6.15)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn>=0.20.1->dash-bio) (1.5.1)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn>=0.20.1->dash-bio) (3.6.0)\n",
            "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.11/dist-packages (from importlib-metadata->dash) (3.23.0)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema->dash-bio) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema->dash-bio) (2025.4.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema->dash-bio) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema->dash-bio) (0.26.0)\n",
            "Requirement already satisfied: pyparsing in /usr/local/lib/python3.11/dist-packages (from periodictable->dash-bio) (3.2.3)\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "ImportError",
          "evalue": "cannot import name 'gc_content' from 'Bio.SeqUtils' (/usr/local/lib/python3.11/dist-packages/Bio/SeqUtils/__init__.py)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-10-3761836930.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mgoogle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolab\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mfiles\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mcollections\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mCounter\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeqUtils\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mProtParam\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgc_content\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mdash_bio\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mdashbio\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mImportError\u001b[0m: cannot import name 'gc_content' from 'Bio.SeqUtils' (/usr/local/lib/python3.11/dist-packages/Bio/SeqUtils/__init__.py)",
            "",
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"
          ],
          "errorDetails": {
            "actions": [
              {
                "action": "open_url",
                "actionText": "Open Examples",
                "url": "/notebooks/snippets/importing_libraries.ipynb"
              }
            ]
          }
        }
      ],
      "source": [
        "# Install required libraries\n",
        "!pip install dash plotly pandas biopython dash-bio\n",
        "import pandas as pd\n",
        "from dash import Dash, dcc, html, dash_table, Input, Output\n",
        "import plotly.express as px\n",
        "import plotly.graph_objects as go\n",
        "from google.colab import files\n",
        "from collections import Counter\n",
        "from Bio.SeqUtils import ProtParam, gc_content\n",
        "import dash_bio as dashbio"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Load Dataset\n",
        "uploaded = files.upload()\n",
        "file_name = next(iter(uploaded))\n",
        "df = pd.read_csv(file_name)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 73
        },
        "id": "eDvbSY0mwEsy",
        "outputId": "5750d1e9-0441-42f2-d301-23fa1357ea34"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-30d59ee2-828e-4371-8eb6-664e592167be\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-30d59ee2-828e-4371-8eb6-664e592167be\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving protein sequence.csv to protein sequence.csv\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Calculate sequence features\n",
        "def analyze_sequence(seq, mol_type):\n",
        "    try:\n",
        "        if mol_type == 'Protein':\n",
        "            analyzer = ProtParam.ProteinAnalysis(seq)\n",
        "            return {\n",
        "                'molecular_weight': analyzer.molecular_weight(),\n",
        "                'aromaticity': analyzer.aromaticity(),\n",
        "                'instability_index': analyzer.instability_index(),\n",
        "                'gravy': analyzer.gravy()\n",
        "            }\n",
        "        else:  # DNA/RNA\n",
        "            return {\n",
        "                'gc_content': GC(seq),\n",
        "                'at_content': 100 - GC(seq)\n",
        "            }\n",
        "    except:\n",
        "        return {}\n"
      ],
      "metadata": {
        "id": "YDyobhyewE8H"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "app = Dash(__name__)"
      ],
      "metadata": {
        "id": "WiiVDH6hzLvo"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Data Preparation\n",
        "sample_df = df.sample(5000) if len(df) > 5000 else df.copy()\n",
        "\n"
      ],
      "metadata": {
        "id": "MbPRYH5fwFdX"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Initialize Dash App\n",
        "app = Dash(__name__)\n"
      ],
      "metadata": {
        "id": "4EGOAXK71Ujp"
      },
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Define Layout\n",
        "app.layout = html.Div([\n",
        "    html.H1(\"Macromolecular Sequence Analysis Dashboard\",\n",
        "           style={'textAlign': 'center', 'color': '#2c3e50', 'marginBottom': '30px'}),\n",
        "\n",
        "    # Row 1: Summary Cards\n",
        "    html.Div([\n",
        "        html.Div([\n",
        "            html.H3(\"Total Sequences\"),\n",
        "            html.P(f\"{len(df):,}\")\n",
        "        ], className='card', style={'background': '#3498db'}),\n",
        "\n",
        "        html.Div([\n",
        "            html.H3(\"Unique Structures\"),\n",
        "            html.P(f\"{df['structureId'].nunique():,}\")\n",
        "        ], className='card', style={'background': '#2ecc71'}),\n",
        "\n",
        "        html.Div([\n",
        "            html.H3(\"Longest Sequence\"),\n",
        "            html.P(f\"{df['seqLength'].max()} residues\")\n",
        "        ], className='card', style={'background': '#e74c3c'}),\n",
        "\n",
        "        html.Div([\n",
        "            html.H3(\"Most Common Type\"),\n",
        "            html.P(\"Protein\" if df['macromoleculeType'].value_counts().index[0] == 'Protein' else \"DNA/RNA\")\n",
        "        ], className='card', style={'background': '#f39c12'})\n",
        "    ], className='card-container'),\n",
        "\n",
        "    # Row 2: Filters\n",
        "    html.Div([\n",
        "        dcc.Dropdown(\n",
        "            id='mol-type-filter',\n",
        "            options=[{'label': mol_type, 'value': mol_type}\n",
        "                    for mol_type in df['macromoleculeType'].unique()],\n",
        "            value='Protein',\n",
        "            style={'width': '30%', 'display': 'inline-block'}\n",
        "        ),\n",
        "        dcc.RangeSlider(\n",
        "            id='length-slider',\n",
        "            min=df['seqLength'].min(),\n",
        "            max=min(df['seqLength'].max(), 1000),  # Cap at 1000 for usability\n",
        "            value=[100, 300],\n",
        "            marks={i: str(i) for i in range(0, 1001, 100)},\n",
        "            step=10,\n",
        "            tooltip={'placement': 'bottom'}\n",
        "        )\n",
        "    ], style={'padding': '20px'}),\n",
        "\n",
        "    # Row 3: Visualizations\n",
        "    html.Div([\n",
        "        dcc.Graph(id='length-distribution'),\n",
        "        dcc.Graph(id='composition-heatmap')\n",
        "    ], style={'columnCount': 2}),\n",
        "\n",
        "    # Row 4: Sequence Analysis\n",
        "    html.Div([\n",
        "        dcc.Dropdown(\n",
        "            id='sequence-selector',\n",
        "            options=[{'label': f\"{row['structureId']}_{row['chainId']} ({row['seqLength']}aa)\",\n",
        "                     'value': i}\n",
        "                    for i, row in sample_df.iterrows()],\n",
        "            value=0,\n",
        "            style={'width': '100%'}\n",
        "        ),\n",
        "        html.Div([\n",
        "            html.H4(\"Sequence Preview:\"),\n",
        "            html.Pre(id='sequence-display', style={\n",
        "                'overflowX': 'auto',\n",
        "                'backgroundColor': '#f5f5f5',\n",
        "                'padding': '10px',\n",
        "                'borderRadius': '5px',\n",
        "                'maxHeight': '200px'\n",
        "            })\n",
        "        ]),\n",
        "        html.Div(id='sequence-stats')\n",
        "    ], style={'border': '1px solid #ddd', 'padding': '20px', 'margin': '20px 0'}),\n",
        "\n",
        "    # Data Table\n",
        "    dash_table.DataTable(\n",
        "        id='data-table',\n",
        "        columns=[{'name': col, 'id': col} for col in df.columns],\n",
        "        data=sample_df.head(10).to_dict('records'),\n",
        "        page_size=10,\n",
        "        style_table={'overflowX': 'auto', 'height': '300px'},\n",
        "        style_cell={\n",
        "            'textAlign': 'left',\n",
        "            'padding': '8px',\n",
        "            'whiteSpace': 'normal',\n",
        "            'maxWidth': '200px'\n",
        "        },\n",
        "        filter_action='native'\n",
        "    )\n",
        "], style={'fontFamily': 'Arial', 'padding': '20px'})\n",
        "\n"
      ],
      "metadata": {
        "id": "5t03zQ4W1Uq0"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Callbacks\n",
        "@app.callback(\n",
        "    [Output('length-distribution', 'figure'),\n",
        "     Output('composition-heatmap', 'figure')],\n",
        "    [Input('mol-type-filter', 'value'),\n",
        "     Input('length-slider', 'value')]\n",
        ")\n",
        "def update_visualizations(mol_type, length_range):\n",
        "    filtered_df = df[(df['macromoleculeType'] == mol_type) &\n",
        "                    (df['seqLength'] >= length_range[0]) &\n",
        "                    (df['seqLength'] <= length_range[1])]\n",
        "\n",
        "    # Length distribution plot\n",
        "    length_fig = px.histogram(filtered_df, x='seqLength', nbins=50,\n",
        "                            title=f'{mol_type} Sequence Length Distribution',\n",
        "                            color_discrete_sequence=['#3498db'])\n",
        "\n",
        "    # Composition heatmap (sample 100 sequences)\n",
        "    sample = filtered_df.sample(min(100, len(filtered_df)))\n",
        "    if mol_type == 'Protein':\n",
        "        aa_counts = pd.DataFrame([Counter(seq) for seq in sample['sequence']]).mean().reset_index()\n",
        "        aa_counts.columns = ['Amino Acid', 'Frequency']\n",
        "        heatmap_fig = px.bar(aa_counts, x='Amino Acid', y='Frequency',\n",
        "                            title='Amino Acid Composition',\n",
        "                            color='Frequency',\n",
        "                            color_continuous_scale='Viridis')\n",
        "    else:\n",
        "        nt_counts = pd.DataFrame([Counter(seq) for seq in sample['sequence']]).mean().reset_index()\n",
        "        nt_counts.columns = ['Nucleotide', 'Frequency']\n",
        "        heatmap_fig = px.bar(nt_counts, x='Nucleotide', y='Frequency',\n",
        "                            title='Nucleotide Composition',\n",
        "                            color='Frequency',\n",
        "                            color_continuous_scale='Viridis')\n",
        "\n",
        "    return length_fig, heatmap_fig\n",
        "\n"
      ],
      "metadata": {
        "id": "9TUnQyGo1UuJ"
      },
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "@app.callback(\n",
        "    [Output('sequence-display', 'children'),\n",
        "     Output('sequence-stats', 'children')],\n",
        "    [Input('sequence-selector', 'value')]\n",
        ")\n",
        "def update_sequence_viewer(selected_idx):\n",
        "    row = sample_df.iloc[selected_idx]\n",
        "\n",
        "    # Calculate sequence statistics\n",
        "    def analyze_sequence(seq, mol_type):\n",
        "        try:\n",
        "            if mol_type == 'Protein':\n",
        "                analyzer = ProtParam.ProteinAnalysis(seq)\n",
        "                return {\n",
        "                    'molecular_weight': analyzer.molecular_weight(),\n",
        "                    'aromaticity': analyzer.aromaticity(),\n",
        "                    'instability_index': analyzer.instability_index(),\n",
        "                    'gravy': analyzer.gravy()\n",
        "                }\n",
        "            else:  # DNA/RNA\n",
        "                return {\n",
        "                    'gc_content': GC(seq),\n",
        "                    'at_content': 100 - GC(seq)\n",
        "                }\n",
        "        except:\n",
        "            return {}\n",
        "\n",
        "    stats = analyze_sequence(row['sequence'], row['macromoleculeType'])\n",
        "\n",
        "    # Display first 200 characters of sequence\n",
        "    seq_display = row['sequence'][:200] + ('...' if len(row['sequence']) > 200 else '')\n",
        "\n",
        "    # Create stats display\n",
        "    if row['macromoleculeType'] == 'Protein':\n",
        "        stats_html = html.Div([\n",
        "            html.H4(f\"{row['structureId']}_{row['chainId']} Protein Analysis\"),\n",
        "            html.P(f\"Length: {row['seqLength']} residues\"),\n",
        "            html.P(f\"Molecular Weight: {stats.get('molecular_weight', 'N/A'):.1f} Da\"),\n",
        "            html.P(f\"Aromaticity: {stats.get('aromaticity', 'N/A'):.2f}\"),\n",
        "            html.P(f\"Instability Index: {stats.get('instability_index', 'N/A'):.1f}\"),\n",
        "            html.P(f\"GRAVY: {stats.get('gravy', 'N/A'):.2f}\")\n",
        "        ])\n",
        "    else:\n",
        "        stats_html = html.Div([\n",
        "            html.H4(f\"{row['structureId']}_{row['chainId']} DNA/RNA Analysis\"),\n",
        "            html.P(f\"Length: {row['seqLength']} bases\"),\n",
        "            html.P(f\"GC Content: {stats.get('gc_content', 'N/A'):.1f}%\"),\n",
        "            html.P(f\"AT Content: {stats.get('at_content', 'N/A'):.1f}%\")\n",
        "        ])\n",
        "\n",
        "    return seq_display, stats_html\n",
        "\n"
      ],
      "metadata": {
        "id": "xi6lcbjr11Iw"
      },
      "execution_count": 17,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Run App\n",
        "app.run(mode='inline')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 671
        },
        "id": "Gc5lWduI11Qu",
        "outputId": "4aefe723-617e-42fd-cb3f-ca14fb1036fb"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "(async (port, path, width, height, cache, element) => {\n",
              "    if (!google.colab.kernel.accessAllowed && !cache) {\n",
              "      return;\n",
              "    }\n",
              "    element.appendChild(document.createTextNode(''));\n",
              "    const url = await google.colab.kernel.proxyPort(port, {cache});\n",
              "    const iframe = document.createElement('iframe');\n",
              "    iframe.src = new URL(path, url).toString();\n",
              "    iframe.height = height;\n",
              "    iframe.width = width;\n",
              "    iframe.style.border = 0;\n",
              "    iframe.allow = [\n",
              "        'accelerometer',\n",
              "        'autoplay',\n",
              "        'camera',\n",
              "        'clipboard-read',\n",
              "        'clipboard-write',\n",
              "        'gyroscope',\n",
              "        'magnetometer',\n",
              "        'microphone',\n",
              "        'serial',\n",
              "        'usb',\n",
              "        'xr-spatial-tracking',\n",
              "    ].join('; ');\n",
              "    element.appendChild(iframe);\n",
              "  })(8050, \"/\", \"100%\", 650, false, window.element)"
            ]
          },
          "metadata": {}
        }
      ]
    }
  ]
}